### PR TITLE
doc: dev add a note about ccache

### PR DESCRIPTION
**Pull request by theanalyst**
Signed-off-by: Abhishek Lekshmanan <abhishek@suse.com>

### CODE REVIEW

File: doc/dev/index.rst [hunk 1]
```diff
@@ -165,6 +165,29 @@ Building from source
 
 See instructions at :doc:`/install/build-ceph`.
 
+Using ccache to speed up local builds
+-------------------------------------
+
+`ccache`_ can speed up rebuilds significantly, since `make clean` is a common
```
Comment (smithfarm) [inline doc/dev/index.rst]:
```
s/`make clean`/``make clean``/
```

File: doc/dev/index.rst [hunk 2]
```diff
@@ -165,6 +165,29 @@ Building from source
...
+enough developer practice. ccache is available as a package in most distros.
+It is recommended to set symlinks to ccache to your compiler for speeding up
+all builds. This can be done by symlinking ccache to the compiler, assuming
+ccache is installed & in `/usr/bin/ccache` ::
```
Comment (smithfarm) [inline doc/dev/index.rst]:
```
s/`/usr/bin/ccache`/``/usr/bin/ccache``/
```

File: doc/dev/index.rst (hunk 1 above)
Comment (theanalyst) [inline doc/dev/index.rst]: ah the markdown -> rst stuff, there are few others too in the pr, I'll get this

File: doc/dev/index.rst [hunk 3]
```diff
@@ -165,6 +165,29 @@ Building from source
...
+
+  $ ln -s /usr/bin/ccache /usr/local/bin/gcc
+  $ ln -s /usr/bin/ccache /usr/local/bin/g++
+  $ ln -s /usr/bin/ccache /usr/local/bin/cc
+  $ ln -s /usr/bin/ccache /usr/local/bin/c++
+
+for more details refer to the `run modes`_ in the ccache manual. The default
+settings of `ccache` can be displayed with `ccache -s`. It is recommended to
```
Comment (smithfarm) [inline doc/dev/index.rst]:
```
s/`ccache`/``ccache``/
```

File: doc/dev/index.rst (hunk 1 above)
Comment (smithfarm) [inline doc/dev/index.rst]: also, "so" or "hence" would be better than "since" in this context

File: doc/dev/index.rst (hunk 2 above)
Comment (smithfarm) [inline doc/dev/index.rst]: can you also change this:

```
OLD: "assuming ccache is installed & in `/usr/bin/ccache`"

NEW: "assuming ccache is installed and the binary is in the standard location `/usr/bin/ccache`
```

File: doc/dev/index.rst [hunk 4]
```diff
@@ -165,6 +165,29 @@ Building from source
...
+It is recommended to set symlinks to ccache to your compiler for speeding up
+all builds. This can be done by symlinking ccache to the compiler, assuming
+ccache is installed & in `/usr/bin/ccache` ::
+
+  $ ln -s /usr/bin/ccache /usr/local/bin/gcc
+  $ ln -s /usr/bin/ccache /usr/local/bin/g++
+  $ ln -s /usr/bin/ccache /usr/local/bin/cc
+  $ ln -s /usr/bin/ccache /usr/local/bin/c++
+
+for more details refer to the `run modes`_ in the ccache manual. The default
```
Comment (smithfarm) [inline doc/dev/index.rst]: the "run modes" section of the ccache manual

**Comment by smithfarm** [comment]
@theanalyst Thanks, looks great. Just some minor nits.

File: doc/dev/index.rst (hunk 1 above)
Comment (theanalyst) [inline doc/dev/index.rst]: I thought I'll mention how to do clean rebuilds as well
See this:

```rst
Rebuilds of the ceph source tree can benefit significantly by using `ccache`_.
Many a times while switching branches and such, one might see build failures for
certain older projects mostly due to older build artifacts. These rebuilds can
significantly benefit the use of ccache. For a full clean source tree, one could
do ::

  $ make clean

  # note the following will nuke everything in the source tree that
  # isn't tracked by git, so make sure to backup any log files /conf options

  $ git clean -fdx; git submodule foreach git clean -fdx

```

Comment (theanalyst) [inline doc/dev/index.rst]: s/projects/branches

File: doc/dev/index.rst (hunk 4 above)
Comment (theanalyst) [inline doc/dev/index.rst]: ack

File: doc/dev/index.rst (hunk 3 above)
Comment (theanalyst) [inline doc/dev/index.rst]: ack

File: doc/dev/index.rst (hunk 2 above)
Comment (theanalyst) [inline doc/dev/index.rst]: ack

File: doc/dev/index.rst (hunk 1 above)
Comment (smithfarm) [inline doc/dev/index.rst]: @theanalyst Looks good; just one nit: "These rebuilds can significantly benefit **from** the use of ccache."

Comment (theanalyst) [inline doc/dev/index.rst]: ack

File: doc/dev/index.rst [hunk 5]
```diff
@@ -165,6 +165,41 @@ Building from source
...
+  # isn't tracked by git, so make sure to backup any log files /conf options
+
+  $ git clean -fdx; git submodule foreach git clean -fdx
+
+ccache is available as a package in most distros. It is recommended to set
+symlinks to ccache to your compiler for speeding up all builds. This can be done
+by symlinking ccache to the compiler, assuming ccache is installed and the
+binary is in the standard location  ``/usr/bin/ccache`` ::
+
+  $ ln -s /usr/bin/ccache /usr/local/bin/gcc
```
Comment (tchaikov) [inline doc/dev/index.rst]: to build ceph with ccache, one can:
```
cmake -DWITH_CCACHE=ON ..
```

to enable the cmake build without pointing gcc/g++/cc/c++ to ccache.

Comment (smithfarm) [inline doc/dev/index.rst]: @tchaikov That's a cmake-only feature, though, right? Not available for autotools builds?

Comment (tchaikov) [inline doc/dev/index.rst]: yes. for automake, one can just use `make CC='ccache gcc' CXX='ccache g++'`, that's a less intrusive way which works even if you don't have the root access of the building host.

Comment (theanalyst) [inline doc/dev/index.rst]: hmm alright I'll add these options as well

Comment (liewegas) [inline doc/dev/index.rst]: I'd just do the cmake instructions since autotools is gone now

Comment (theanalyst) [inline doc/dev/index.rst]: updated

File: doc/dev/index.rst [hunk 6]
```diff
@@ -165,6 +165,47 @@ Building from source
...
+  # isn't tracked by git, so make sure to backup any log files /conf options
+
+  $ git clean -fdx; git submodule foreach git clean -fdx
+
+ccache is available as a package in most distros. To build ceph with ccache one
+can::
+
+  $ cmake -DWITH_CCACHE=ON ..
+
+ccache can also be used for speeding up all builds in the system. This can be
```
Comment (tchaikov) [inline doc/dev/index.rst]: it's a nice introduction to ccache. but i think what we need is doc specific to ceph.

Comment (theanalyst) [inline doc/dev/index.rst]: I guess I can just point to the run-modes link and drop the commands etc.

Comment (theanalyst) [inline doc/dev/index.rst]: dropped the section

**Comment by tchaikov** [comment]
@smithfarm are your comments addressed in the latest change?

**Comment by smithfarm** [comment]
Jenkins re-test this please

**Pull request merged**